core/src/main/scala/cats/data/HashMap.scala

```diff
@@ -0,0 +1,1211 @@
+      *
+      * @param bitMap the bitmap indicating either data value or node positions in the contents array.
+      * @param bitPos the notional index of the element in the trie node.
+      * @return the absolute index of an element in the contents array.
+      */
+    final def indexFrom(bitMap: Int, bitPos: Int): Int =
+      Integer.bitCount(bitMap & (bitPos - 1))
+
+    /**
+      * Creates a new empty bitmap node.
+      *
+      * @param hash the [[cats.kernel.Hash]] instance to use to hash elements.
+      * @return a new empty bitmap node.
+      */
+    final def empty[K, V](implicit hashKey: Hash[K]): Node[K, V] =
+      new BitMapNode[K, V](0, 0, Array.empty[Any], 0)
+  }
+
+  private[HashMap] class Iterator[K, V] extends scala.collection.AbstractIterator[(K, V)] {
+    private var currentNode: Node[K, V] = null
+
+    private var currentValuesIndex: Int = 0
+    private var currentValuesLength: Int = 0
+
+    private var currentDepth: Int = -1
+
+    private val nodeStack: Array[Node[K, V]] =
+      new Array(Node.MaxDepth)
+
+    private val nodeIndicesAndLengths: Array[Int] =
+      new Array(Node.MaxDepth * 2)
+
+    def this(rootNode: Node[K, V]) = {
+      this()
+      if (rootNode.hasNodes) pushNode(rootNode)
+      if (rootNode.hasKeyValues) pushValues(rootNode)
+    }
+
+    final private def pushNode(node: Node[K, V]): Unit = {
+      currentDepth += 1
+
+      val cursorIndex = currentDepth * 2
+      val lengthIndex = currentDepth * 2 + 1
+
+      nodeStack(currentDepth) = node
+
+      nodeIndicesAndLengths(cursorIndex) = 0
+      nodeIndicesAndLengths(lengthIndex) = node.nodeCount
+    }
+
+    final private def pushValues(node: Node[K, V]): Unit = {
+      currentNode = node
+      currentValuesIndex = 0
+      currentValuesLength = node.keyValueCount
+    }
+
+    final private def getMoreValues(): Boolean = {
+      var foundMoreValues = false
+
+      while (!foundMoreValues && currentDepth >= 0) {
+        val cursorIndex = currentDepth * 2
+        val lengthIndex = currentDepth * 2 + 1
+
+        val nodeIndex = nodeIndicesAndLengths(cursorIndex)
+        val nodeLength = nodeIndicesAndLengths(lengthIndex)
+
+        if (nodeIndex < nodeLength) {
+          val nextNode = nodeStack(currentDepth)
+            .getNode(nodeIndex)
+
+          if (nextNode.hasNodes) {
+            pushNode(nextNode)
+          }
+
+          if (nextNode.hasKeyValues) {
+            pushValues(nextNode)
+            foundMoreValues = true
+          }
+
+          nodeIndicesAndLengths(cursorIndex) += 1
+
+        } else {
+          currentDepth -= 1
+        }
+      }
+
+      foundMoreValues
+    }
+
+    final override def hasNext: Boolean =
+      (currentValuesIndex < currentValuesLength) || getMoreValues()
+
+    final override def next(): (K, V) = {
+      if (!hasNext) throw new NoSuchElementException
+      val value = currentNode.getMapping(currentValuesIndex)
```

Comment:
I think this is literally the only place where we don't destructure after getting the key and value.

Probably better to just allocate the tuple here where it actually needed.